File: 304.py
from typing import List
# class NumMatrix:
#     # 通过一维进行二维的拼接 一维的计算方式看leetcode 303题
#     def __init__(self, matrix: List[List[int]]):
#         row = len(matrix)
#         col = len(matrix[0])
#         d = [[0] for _ in range(row)]
#         for a in range(row):
#             for b in range(col):
#                 d[a].append(d[a][-1]+matrix[a][b])
#         self.d = d


#     def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
#         res = 0
#         for a in range(row1,row2+1):
#             res += self.d[a][col2+1]-self.d[a][col1]
#         return res
class NumMatrix:
    # 直接进行二维的计算
    def __init__(self, matrix: List[List[int]]):
        if not matrix or not matrix[0]:
            row , col  = 0,0
        else:
            row ,col = len(matrix),len(matrix[0])
        self.preSum = [[0]*(col+1) for _ in range(row+1)]
        for  a in range(row):
            for b in range(col):
                self.preSum[a+1][b+1]=self.preSum[a][b+1]+self.preSum[a+1][b]-self.preSum[a][b]+matrix[a][b]


    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        return self.preSum[row2+1][col2+1]-self.preSum[row2+1][col1]-self.preSum[row1][col2+1]+self.preSum[row1][col1]
s = NumMatrix([
  [3, 0, 1, 4, 2],
  [5, 6, 3, 2, 1],
  [1, 2, 0, 1, 5],
  [4, 1, 0, 1, 7],
  [1, 0, 3, 0, 5]
])
print(s.sumRegion(1,1,2,2))

File: leetcode_c_plus_plus/304.cpp
using namespace std;
#include<vector>
class NumMatrix {
public:
    // 傻逼代码 跟官方一样结果超时
    vector<vector<int>> sums ;
    NumMatrix(vector<vector<int>>& matrix) {
        int row = matrix.size();
        if(row > 0){
            int col = matrix[0].size();
            sums.resize(row+1,vector<int>(col+1));
            for(int i = 1;i < row+1;i++){
                for(int j = 1;j < col+1;j++){
                    sums[i][j] = sums[i-1][j]+sums[i][j-1]+matrix[i-1][j-1]-sums[i-1][j-1];
                }
            }
        }
        
    }

    
    int sumRegion(int row1, int col1, int row2, int col2) {
        return sums[row2 + 1][col2 + 1] - sums[row1][col2 + 1] - sums[row2 + 1][col1] + sums[row1][col1];
    }
};

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix* obj = new NumMatrix(matrix);
 * int param_1 = obj->sumRegion(row1,col1,row2,col2);
 */